### DISCUSSION TITLE
An ECHO bug in the telnet protocol?

### DISCUSSION THREAD

**Discussion by cfcohen** [evennia/evennia#2776] [Evennia Questions]
This is probably a bug, but I'm having a hard time nailing down the fix, so I've decided to report it as a question in case I'm just doing something wrong.   This is a standard Evennia new game [repo](https://github.com/cfcohen/evennia-game/commit/c5016d48ed6ad7b0cedd0a39b374146a2085ff01) with one minimal commit that overrides the login command set with an EvMenu to prompt for a username and password.  It's very similar to several examples in the documentation and the contrib section, except that it attempts to disable local echo for the password using the telnet ECHO feature.

And it works.  Mostly.  It logs in correctly, it hides the passwords correctly, and there are no obvious problems.   Except that if you run the Mudlet client so that the telnet debugging messages go to standard out (the default configuration in my build), under some circumstances it spews _constantly_.    Specifically, everything works perfectly (no spew) if you enter your password correctly on the first attempt.   But if you enter the password incorrectly the first time, and correctly on the second attempt, the EvMenu logic that sends you back to node_password node somehow causes a tight client-server loop sending packets that result in these debug messages about a dozen times per second:

```
Server sent telnet IAC WILL ECHO (1)
WE send telnet IAC DO ECHO (1)
Enabling Server ECHOing of our output - perhaps he want us to type a password?
Server sent telnet IAC WONT ECHO (1)
WE send telnet IAC DONT ECHO (1)
Server is stopping the ECHOing our output - so back to normal after, perhaps, sending a password...
```

What's particularly annoying about the problem is that it persists for the entire duration of the connection, and everything _seems_ fine, but there's all this extra traffic making it quite slow.   The "bug" is eliminated by sending a failed password attempt to node_bogus from node_password, and then back to node_password again, which seems like a very curious fix to me.  I've confirmed that evennia.server.portal.TelnetProtocol.disableLocal() is being called a dozen times per second with the "echo" option.   Apparently, it is not code in Evennia that's sending the responses, but something from twisted.conch.Telnet in response to disableLocal(). I've had some trouble _demonstrating_ the problem in other clients because of a lack of debugging of telnet option commands, but I suspect that the problem is not specific to the Mudlet client.  In particular, it takes two to tango and the inconsistencies scream that it's on the Evennia side to me. :-)

I've also included an "echotest" command in the game that demonstrates that the approach of calling session.msg("...", options={'echo': True) works correctly once the player is logged in.   I've convinced myself that the problem is _not_ related to disabling echo in the node_password() method itself, but the code I used to reach that conclusion involved a bunch of hacky overrides to EvMenu to enable and disable the echo as close to sending the password prompt and receiving the password as I could manage.   Since that code didn't add anything meaningful to the problem description, I have not included it in the repo. 

In summary, my current suspicion is that maybe something is getting out-of-sync in the portal.  The bug could be race-condition related I guess.   It also seems very suspicious to me that enableLocal() and disableLocal() are supposed to be enabling and disabling local options, but instead just return True for echo.   This would make sense if the attitude was that echo was enabled "by default", but what happens if I change it temporarily (e.g. for a password)?  If you have trouble reproducing the problem, putting a print statement in disableLocal() and watching the log has worked for me in addition to the Mudlet spew.

I would be very grateful if someone else can confirm that this bug is not just on my system, that the diff to the code looks reasonable, and if it's truly a bug, I'm happy to open a ticket for it.   Sadly after several hours of trying to fix this issue, I've reached the end of my abilities -- I'm still new to Evennia. :-(

**Comment by cfcohen** [comment]
While reproducing this cleanly one more time from my post above, I discovered that I had (again) mixed build directories from 0.9.5 and 1.0-dev, and thus had a diff against a game directory that was pretty non-standard so I replaced the repo with a clean one built against the clean 0.9.5 template.   Hopefully, nobody saw my mistake. :-)

I also found another "fix" for the problem.  Don't turn echoing back on until the player has successfully authenticated.  I don't think this really fixes what's broken in the tight feedback loop, but it does seem to work around it more gracefully than the node_bogus fix.  I presume this fix somehow works because it avoids turning echo off and on, but it doesn't seem very defensive.  It could leave the connection in a non-echoing state, which might be pretty annoying.

**Comment by Griatch** [comment]
@cfcohen Thanks for the analysis - also great that you supply test code - very useful for reproducing, kudos! :+1: 

I don't see any issues or slowdowns using your code on `tintin++` but it's also possible that the `TELNET DEBUG` mode of `tintin++` is not showing the individual echo handshakes. I'll test with mudlet later to see if I can reproduce.

**Reply by cfcohen** [reply]
I have confirmed (using Wireshark) that the problem does _not_ occur in tintin++ or the Javascript client.  Curious.  Maybe it is a bug in Mudlet?  Since such things may turn out to be important, I'm also running Mudlet 4.16.0-dev, build from source at commit bd91aa87bf40.

**Reply by Griatch** [reply]
I seem able to reproduce this also in the much older Mudlet 3.7.1. Making some probing changes in the telnet protocol didn't have any effect; it seems mudlet keeps sending the disable option over and over. It's unclear where the bug is here - in Mudlet or in Evennia. Since it's not reproducible in tintin++, it could come down to some interpretation of the protocol (it has happened before); it would seem Mudlet is not happy about the response Evennia gives here.

**Comment by cfcohen** [comment]
Thank you for confirming that this isn't just my system configuration, a bad build of Mudlet, my stupidity, or something else.   I'm planning on doing some work in Mudlet with scripts, GMCP, and building a GUI package for Evennia, so I'd like to try and track this down.  In fact, I've tried already and failed. :-(

Since it's hopefully not just blathering at this point, I'll expound on the analysis I've done and the (possibly incorrect) conclusions that I've reached.   The fact that adding a bogus node to the menu seems to fix the problem suggests to me that the issue is corrected by putting some non-control bytes on the wire between the out-of-band IAC ECHO commands.   Since I'm having some trouble understanding why that would be relevant to the bug, my theory is that it's a "state-machine" problem and that the extra traffic moves us out of the broken state machine cycle and back into the intended state machine cycle.

I originally thought that the "state-machine problem" would be somewhere in Evennia, but I now think this is wrapped up in the logic in enableLocal() and disableLocal() in evennia.server.portal.TelnetProtocol.  These functions are the only ones that I've been able to instrument in the Evennia code to print something during the tight ECHO cycle.   Neither of these methods actually send any traffic, and so it's clearly(?) NOT Evennia that's sending the ECHO commands, but instead something in twisted.conch.Telnet in response to what Evennia has returned in those functions.

My reading of the twisted documentation is that those functions are _supposed_ to be _changing_ the state on the local implementation.  Instead they seem to just be giving hard-coded answers without actually modifying the state.   And this is where I'm really lost...  Why isn't there code in telnet.py disableLocal() that looks more like:
```
if option == ECHO:
    self.echo_enabled = False
    return True
```
presumably with corresponding code in the enableLocal() that sometimes returns True and sometimes returns False based on the current status of self.echo_enabled?    I'm not really familiar with twisted, or the Evennia portal (yet?), but I have read the telnet RFC in some detail, and this doesn't feel right to me.   But from my perspective, there are several big problems with my theories and my mucking around in this code -- 1. It clearly works most of the time for most options, including ECHO.  2. I clearly don't understand _how_ it works most of the time.  3. It seems very unwise of me to try to propose fixes in this part of the code given how sensitive it is to breaking other clients and so forth.

That's why I finally opened a ticket on this.  If you can reproduce the behavior in disableLocal(), and say something about how that function is _supposed_ to be working, I might be able to make another set of tests trying to figure out when and why it's not functioning as intended.

Thanks again for your help.  I suppose in the big picture, this is a relatively niche problem (Mudlet+specific ECHO requirements+reluctance to just work around it).  But the fact that Evennia seems to have sorted out many of the server-client protocol issues, and that it has implemented this nice separate portal concept is one of the stronger points of Evennia to me, so I feel like really getting those issues sorted is very important.  Similarly, I've noticed some inconsistent behavior involving the translation of true and false values between Mudlet and Evennia over GMCP and MSDP, but I'll save that for a separate discussion topic.
